### PR TITLE
drivers/feetech : initial support

### PR DESCRIPTION
**Pull request by astralien3000**
Hello !

I'm adding a driver for feetech's scs15 servomotors.
I'm basing my work on [my previous C++ library](https://github.com/AversivePlusPlus/feetech).

This is currently a work in progress.

EDIT : 
[Here is](http://www.feetechrc.com/product/digital-servo/feetech-scservo-multi-purpose-uart-bus-robot-servo-for-arduino-diy/) the best source of information I can find on the motors .

### CODE REVIEW

File: drivers/feetech/include/feetech.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#define FEETECH_STREAM_TYPE uart_t
+#define FEETECH_STREAM_INIT(stream) uart_init(stream, 1000000, 0, 0) // TODO
+#define FEETECH_STREAM_WRITE(stream, buffer, size) uart_write(stream, buffer, size)
+#define FEETECH_STREAM_READ(stream, buffer, size) ((void)buffer, size) // TODO
+
+typedef FEETECH_STREAM_TYPE feetech_stream_t;
+typedef uint8_t feetech_device_t;
+
+void feetech_init(feetech_stream_t stream) {
+  FEETECH_STREAM_INIT(stream);
```
Comment (aabadie) [inline drivers/feetech/include/feetech.h]: indent is wrong, it should be 4 characters wide

**Comment by aabadie** [comment]
Nice ! 
I had a brief look at your code : it would be better if you could provide a clean API in `drivers/include/feetech.h` and move the internal implementation + constants in `drivers/feetech`. Here you put all code in `drivers/feetech/include` with all implementation in header files.

File: drivers/feetech/include/feetech.h (hunk 1 above)
Comment (astralien3000) [inline drivers/feetech/include/feetech.h]: OK.
I did not follow the coding conventions for the moment because I just translated from C++.
It will be done !

**Comment by astralien3000** [comment]
Sure ! It is the plan.
I should also move the code from headers to source files.

File: drivers/feetech/feetech.c [hunk 2]
```diff
@@ -1,3 +1,11 @@
+/*
+ * Copyright (C) 2017 INRIA
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: You may want to had some doxygen comment here:
```
/**
 * @ingroup     drivers_feetech
 * @{
 *
 * @file
 * @brief       Device driver implementation for Feetech devices
 *
 * @author      Loïc Dauphin <loic.dauphin@inria.fr>
 *
 * @}
 */
```

Comment (astralien3000) [inline drivers/feetech/feetech.c]: Thanks !

**Comment by astralien3000** [comment]
Looks like feetech.h is well documented considering Jenkins's test.

questions : 
 - for `bool` returning function, would it be a better idea to return `int` (0 on success, <0 on error) ?
 - for `size_t` returning function, i used the "POSIX stream" convention, but it has no sense to return an value other than 0 or `length`, should I use the previous point's convention (`bool` or `int`) ?

**Comment by astralien3000** [comment]
I also added a link to a technical page of the servos to the head message of this PR.

**Comment by astralien3000** [comment]
I really don't understand this error : 
```
reader.c: In function 'feetech_reader_check_sum':
reader.c:51:70: error: comparison of promoted ~unsigned with unsigned [-Werror=sign-compare]
     return ((uint8_t)(~(uint8_t)feetech_reader_compute_sum(reader))) == ((uint8_t)reader->buffer[reader->size-1]);
```

File: drivers/feetech/feetech.c [hunk 3]
```diff
@@ -29,12 +29,14 @@
...
     uart_write(stream->uart, buffer, size);
     return size;
 }
 
-static size_t _feetech_stream_read(feetech_stream_t *stream, uint8_t *buffer, size_t size) {
+static size_t _feetech_stream_read(feetech_stream_t *stream, uint8_t *buffer, size_t size)
+{
     uint32_t begin = xtimer_now_usec();
     uint32_t cur = begin;
     while(cur-begin < FEETECH_TIMEOUT_MS*1000) {
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: Prefer putting space around arithmetic operators. It seems to be the convention used in other places in the code

**Comment by astralien3000** [comment]
Squashed, I think I am ready for CI, and manual tests (ping @aabadie ? @kYc0o ?)

**Comment by astralien3000** [comment]
Why is Jenkins build so long ?

**Comment by astralien3000** [comment]
new todo list after working on the PR #6725 : 
 - Try to uniformize the interface between feetech and dynamixel
 - Add a way to specify the TX enable/disable
 - Rename enums in `feetech_protocol.h`

File: drivers/Makefile.dep [hunk 4]
```diff
@@ -199,3 +199,9 @@ ifneq (,$(filter xbee,$(USEMODULE)))
   USEMODULE += xtimer
   USEMODULE += netif
 endif
+
+ifneq (,$(filter uart_half_duplex,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_gpio
```
Comment (aabadie) [inline drivers/Makefile.dep]: Indentation is 2 spaces, it should be 4

File: drivers/feetech/feetech.c [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#include "feetech_protocol.h"
+#include "feetech_reader.h"
+#include "feetech_writer.h"
+
+#include "periph/uart.h"
+#include "xtimer.h"
+
+#include <string.h>
+
+static inline void _n2h(uint16_t* value) {
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: `{` should be on another line

File: drivers/feetech/feetech.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+void feetech_init(feetech_t *device, uart_half_duplex_t *stream, feetech_id_t id)
+{
+    device->stream = stream;
+    device->id = id;
+}
+
+bool feetech_write(feetech_t *device, feetech_addr_t reg, const uint8_t *data, size_t length)
+{
+    if(device->stream->size < length) {
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: space missing after `if`. I saw this in lots of other places.

File: drivers/feetech/feetech.c [hunk 7]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2017 INRIA
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: It should be `Inria` instead of `INRIA`

File: drivers/feetech/include/feetech_protocol.h [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+#ifndef FEETECH_PROTOCOL_H
+#define FEETECH_PROTOCOL_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define FEETECH_START ((uint8_t)0xFF)
```
Comment (aabadie) [inline drivers/feetech/include/feetech_protocol.h]: I think the cast is not needed

File: drivers/feetech/include/feetech_protocol.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    SCS15_B_0_5M   = 1,
+    SCS15_B_250K   = 2,
+    SCS15_B_128K   = 3,
+    SCS15_B_115200 = 4,
+    SCS15_B_76800  = 5,
+    SCS15_B_57600  = 6,
+    SCS15_B_38400  = 7
+} scs15_baudrate_t;
+
+//8bits register Address
```
Comment (aabadie) [inline drivers/feetech/include/feetech_protocol.h]: cpp style comments, please use C style comments (`/* */`)

File: drivers/Makefile.dep (hunk 4 above)
Comment (astralien3000) [inline drivers/Makefile.dep]: Ok, but why is half of this file 2 spaced ?

Comment (aabadie) [inline drivers/Makefile.dep]: indeed, there are inconsistencies about indentation. Maybe that should deserve a dedicated PR (make indentation consistent in the whole Makefile.dep)

File: drivers/feetech/writer.c [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+size_t feetech_writer_get_size(const feetech_writer_t *writer)
+{
+    return writer->size;
+}
+
+void feetech_writer_response_make(feetech_writer_t *writer, uint8_t id, const uint8_t *buffer, size_t size)
+{
+    const size_t len = 2+size;
```
Comment (aabadie) [inline drivers/feetech/writer.c]: missing space around operator

File: drivers/feetech/writer.c [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+
+        writer->buffer[0] = FEETECH_START;
+        writer->buffer[1] = FEETECH_START;
+        sum += writer->buffer[2] = id;
+        sum += writer->buffer[3] = len;
+        sum += writer->buffer[4] = INST_WRITE;
+
+        sum += writer->buffer[5] = reg;
+        for(size_t i = 0 ; i < size ; i++) {
+            sum += writer->buffer[6+i] = buffer[i];
```
Comment (aabadie) [inline drivers/feetech/writer.c]: space around operator

File: drivers/feetech/writer.c [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+        }
+
+        writer->buffer[writer->size-1] = ~sum;
```
Comment (aabadie) [inline drivers/feetech/writer.c]: space

File: drivers/feetech/writer.c [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+{
+    if (writer->size <= 8) {
+        writer->size = 0;
+    }
+}
+
+void feetech_writer_sync_write_add(feetech_writer_t *writer, uint8_t id, const uint8_t *buffer, size_t size)
+{
+    if (feetech_writer_sync_write_required(writer) == size &&
+       size != 0 && writer->size+size+1 <= writer->limit) {
```
Comment (aabadie) [inline drivers/feetech/writer.c]: space

Comment (aabadie) [inline drivers/feetech/writer.c]: indentation is off

File: tests/driver_feetech/main.c [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+             "\t 115200\n"
+             "\t  76800\n"
+             "\t  57600\n"
+             "\t  38400"
+             );
+        return 1;
+    }
+    /* parse parameters */
+    uart = parse_uart(argv[1]);
+    if (uart < 0) return -1;
```
Comment (aabadie) [inline tests/driver_feetech/main.c]: split in 2 ilnes with braces

**Comment by astralien3000** [comment]
Comments addressed

File: drivers/uart_half_duplex/uart_half_duplex.c [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    _enable_tx(dev->dir);
+    uart_write(dev->uart, dev->buffer, size);
+    _disable_tx(dev->dir);
+    return size;
+}
+
+size_t uart_half_duplex_recv(uart_half_duplex_t *dev, size_t size)
+{
+    const uint32_t begin = xtimer_now_usec();
+    while(xtimer_now_usec()-begin < dev->timeout_us) {
```
Comment (aabadie) [inline drivers/uart_half_duplex/uart_half_duplex.c]: Space missing after while

File: tests/driver_feetech/main.c [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+}
+
+static void parse_reg(char *arg, int *reg8, int *reg16)
+{
+    *reg8 = -1;
+    *reg16 = -1;
+
+#define PARSE_REG(var, reg) \
+    if (strcmp(#reg, arg) == 0) { *var = SCS15_##reg; return; }
+    // 8bits
```
Comment (aabadie) [inline tests/driver_feetech/main.c]: cpp comment,  use c style

File: drivers/uart_half_duplex/uart_half_duplex.c (hunk 15 above)
Comment (astralien3000) [inline drivers/uart_half_duplex/uart_half_duplex.c]: the `while` too ?!? I guess I should have done it before... ok

File: tests/driver_feetech/main.c (hunk 16 above)
Comment (astralien3000) [inline tests/driver_feetech/main.c]: sorry, you already had this request for an other comment. I missed this one

**Comment by astralien3000** [comment]
Comment addressed

File: drivers/feetech/feetech.c [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+        return false;
+    }
+
+    feetech_reader_t pr;
+    feetech_reader_init(&pr, device->stream->buffer, esize);
+    if (!feetech_reader_is_valid(&pr)) {
+        return false;
+    }
+
+    if (feetech_reader_response_get_payload_size(&pr) != length) {
```
Comment (aabadie) [inline drivers/feetech/feetech.c]: you are calling the same function twice. Would it make sense to use something like: 
```c
    }
+
+    if (feetech_reader_response_get_payload_size(&pr, data) != length) {
    return false;
}

return true;
```

File: drivers/feetech/include/feetech_protocol.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define FEETECH_START (0xFF)
+
+typedef enum {
+    SCS15_B_1M     = 0,
+    SCS15_B_0_5M   = 1,
```
Comment (aabadie) [inline drivers/feetech/include/feetech_protocol.h]: the numbering is not needed for an enum when they're in order.

File: drivers/feetech/reader.c [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @}
+ */
+
+#include "feetech_reader.h"
+
+static uint8_t _compute_sum(const feetech_reader_t *reader)
+{
+    uint8_t sum = 0;
+    for(size_t i = 2 ; i < reader->size-1 ; i++) {
```
Comment (aabadie) [inline drivers/feetech/reader.c]: space missing

File: drivers/uart_half_duplex/include/uart_half_duplex.h [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+#define UART_HALF_DUPLEX_DIR_NONE           (0)  /**< Don't manage direction */
+
+#define UART_HALF_DUPLEX_DIR_PIN_SET(pin)   (((pin + 1) << 1))     /**< pin set enables TX */
+#define UART_HALF_DUPLEX_DIR_PIN_CLEAR(pin) (((pin + 1) << 1) | 1) /**< pin clear enables TX */
+
+/**
+ * @brief Descriptor struct for half-duplex UART
+ */
+typedef struct {
+    uart_t uart;                 /**< the half-duplex UART bus used */
```
Comment (aabadie) [inline drivers/uart_half_duplex/include/uart_half_duplex.h]: attributes seems duplicated from the params struct, can you replace them a params member ?

File: drivers/feetech/writer.c [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+
+void feetech_writer_sync_write_add(feetech_writer_t *writer, uint8_t id, const uint8_t *buffer, size_t size)
+{
+    if (feetech_writer_sync_write_required(writer) == size &&
+        size != 0 && writer->size + size + 1 <= writer->limit) {
+
+        uint8_t sum = ~writer->buffer[writer->size - 1];
+
+        sum += writer->buffer[writer->size - 1] = id;
+        for(size_t i = 0 ; i < size ; i++) {
```
Comment (aabadie) [inline drivers/feetech/writer.c]: space :)

**Comment by aabadie** [comment]
I see 3 remaining things:
* add a README with the test application
* change the bool return values to more explicit int return codes
* squash

**Comment by aabadie** [comment]
Tested works. Just need squashing, and we are good.

File: drivers/feetech/Makefile [hunk 22]
```diff
@@ -0,0 +1,3 @@
+MODULE = feetech
```
Comment (kaspar030) [inline drivers/feetech/Makefile]: not needed here

File: drivers/feetech/feetech.c [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+#include "feetech_protocol.h"
+#include "feetech_reader.h"
+#include "feetech_writer.h"
+
+#include "periph/uart.h"
+#include "xtimer.h"
+
+#include <string.h>
+
+/* Manages conversion between protocol's endianness and host's endianness */
```
Comment (kaspar030) [inline drivers/feetech/feetech.c]: Why not use byteorder.h?

File: drivers/uart_half_duplex/include/uart_half_duplex.h [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+#include <stdlib.h>
+
+#include "periph/uart.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef UART_HALF_DUPLEX_DEFAULT_TIMEOUT_US
+#define UART_HALF_DUPLEX_DEFAULT_TIMEOUT_US (20000) /**< Default recv timeout (in microseconds) */
```
Comment (kaspar030) [inline drivers/uart_half_duplex/include/uart_half_duplex.h]: ```20000LU``` please

File: tests/driver_feetech/main.c [hunk 25]
```diff
@@ -0,0 +1,402 @@
...
+    else if (UART_DEV(uart) == UART_STDIO_DEV) {
+        printf("Error: The selected UART_DEV(%u) is used for the shell!\n", uart);
+        return -2;
+    }
+    return uart;
+}
+
+static int32_t parse_baud(char *arg)
+{
+#define PARSE_BAUD(baud) \
```
Comment (kaspar030) [inline tests/driver_feetech/main.c]: while somewhat elegant, please find a solution without macros. How about a table?

File: tests/driver_feetech/main.c [hunk 26]
```diff
@@ -0,0 +1,402 @@
...
+    }
+    return dev;
+}
+
+static void parse_reg(char *arg, int *reg8, int *reg16)
+{
+    *reg8 = -1;
+    *reg16 = -1;
+
+#define PARSE_REG(var, reg) \
```
Comment (kaspar030) [inline tests/driver_feetech/main.c]: same here.

File: tests/driver_feetech/main.c [hunk 27]
```diff
@@ -0,0 +1,402 @@
...
+uart_half_duplex_t stream;
+
+static int cmd_init(int argc, char **argv) {
+    int uart = -1;
+    int baud = -1;
+    uint32_t timeout = -1;
+
+    if (argc != 3 && argc != 4) {
+        printf("usage; %s <uart> <baudrate> [<timeout_us>]\n", argv[0]);
+        puts("available baudrates :\n"
```
Comment (kaspar030) [inline tests/driver_feetech/main.c]: you could reuse your baudrate table here. ;)

**Comment by aabadie** [comment]
Strange, you still have conflicts with `Makefile.include`. Can you rebase (check that your local origin/master is up-to-date) ?

File: drivers/include/feetech.h [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+ * @return             FEETECH_BUFFER_TOO_SMALL if buffer is too small for the message
+ * @return             FEETECH_INVALID_MESSAGE if an invalid message was received
+ */
+int feetech_read(feetech_t *device, feetech_addr_t addr, uint8_t *data, size_t length);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif//FEETECH_H
```
Comment (aabadie) [inline drivers/include/feetech.h]: c++ comment, use `/*  */`

File: drivers/uart_half_duplex/uart_half_duplex.c [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+
+static void _rx_cb(void* data, uint8_t c)
+{
+    uart_half_duplex_t *dev = data;
+    if (dev->size < dev->limit) {
+        dev->buffer[dev->size++] = c;
+    }
+}
+
+int uart_half_duplex_init(uart_half_duplex_t *dev, uint8_t *buffer, size_t limit, const uart_half_duplex_params_t *params)
```
Comment (aabadie) [inline drivers/uart_half_duplex/uart_half_duplex.c]: s/limit/max_buffer_size/ ?

File: tests/driver_feetech/main.c [hunk 30]
```diff
@@ -0,0 +1,391 @@
...
+    };
+
+    int ret = uart_half_duplex_init(&stream, feetech_buffer, ARRAY_LEN(feetech_buffer), &params);
+
+    if (argc == 4) {
+        stream.timeout_us = timeout;
+    }
+
+    if (ret == UART_HALF_DUPLEX_NODEV) {
+        puts("Error: invalid UART device given");
```
Comment (aabadie) [inline tests/driver_feetech/main.c]: You should return after an error, otherwise you'll also display a success message.

**Comment by astralien3000** [comment]
@kaspar030 can you check if your comments were properly addressed ?

**Comment by aabadie** [comment]
@astralien3000, can you squash ?

**Comment by astralien3000** [comment]
Done !

**Comment by aabadie** [comment]
Needs squashing still :)

**Comment by astralien3000** [comment]
What is the problem with my squashing ?
I thought I had done it right.
What do you want me to do (exactly) ?

**Comment by aabadie** [comment]
Looks like Murdock agrees on squashing ;)

Typically, a list of commits like that:
1. `drivers/uart_half_duplex: initial support`
2. `drivers/feetech: initial support`
3. `tests/driver_feetech: initial support`

**Comment by astralien3000** [comment]
And of course `./dist/tools/pr_check/pr_check.sh` does not print anything on my computer...

**Comment by aabadie** [comment]
use `make static-tests`

**Comment by kaspar030** [comment]
@astralien3000 In this case, Murdock states that the "NEEDS SQUASHING" github label is still set. That is only checked once at build start.

But I think we all agree with the CI, please go with @aabadie's suggested list of commits. :)

**Comment by astralien3000** [comment]
@kaspar030 Ok, thanks, I was wondering what was going wrong with my computer.

**Comment by astralien3000** [comment]
Yay ! Everything seems ok.
If there is nothing to add : 
Thank you very much @kaspar030 and especially @aabadie for all your comments and your help.

**Comment by aabadie** [comment]
All green, Go !
Thanks @astralien3000 for the good work and patience ;)
